src/aspire/classification/class_selection.py

```diff
@@ -107,11 +138,507 @@ def __init__(self, seed=None):
         """
         self.seed = seed
 
-    def _select(self, n, classes, reflections, distances):
+    def _select(self, classes, reflections, distances):
         """
         Select random `n` classes from the population.
         """
+        # Assign uniform quality.
+        self._quality_scores = np.zeros(self.n)
+
         # Instantiate a random Generator
         rng = np.random.default_rng(self.seed)
         # Generate and return indices for random sample
-        return rng.choice(np.max(classes[:, 0]), size=n, replace=False)
+        # +1 for zero indexing
+        return rng.choice(self.n, size=self.n, replace=False)
+
+
+class ContrastClassSelector(ClassSelector):
+    """
+    Selects top classes based on highest contrast,
+    as estimated by variances of `distances`.
```

Comment:
This is a really good question that I hope I can clear up.  Briefly, I think taking the flattened compressed feature space variance is an _approximation_ of the stack image variance.  I haven't done an actual regression (we could), but comparisons seem to track.

Take some true image to set theoretical `u` and having a neighbor set `I` as `Xi`.   Forget RIR for a moment.  Imagine we are just comparing flattened real images.  The NN table would hold all the `norm(Xi - u_i)`.  So for set of `J` pixels, if our norm was like `sum_j ((Xi_j - u_i_j)^2))` (ignore sqrt for now, comes up later).... looks like we're computing a flattened var per image up to scaling.  Okay, cool.  Now if we more/less believe the views are correctly identified then we can say up to rotations we are looking at the same thing, and so have the same distribution. Ignore rotations for the next few sentences (which are just for intuition).  Under the same distribution, we can sum variances.  Now consider the variance of the mean stacked image `X = sum(X_I)/len(I)` with `Var(X) = sum((X_j -u)^2)/len(J)` .  Expanding into sum, `Var(X) = sum_i(Var(X_i)))` up to the difference between `u` and sampled `u_i`.  Then `Var(X) = sum_i( sum_j ((Xi_j - u_i_j)^2)) ) / len(J)` , again up to `u` sampling error.  I believe that sum to be proportional to what is computed from a flattened var up to dividing by `len(I)`, which doesn't matter for sorting.
 
Really NN distance is the norm of the compressed bispectrum power (in a FB basis) with respect to the base image.    So instead of j pixels, we have j "features".  Its not exactly the same, but if we believe that the compression is working, and Plancherel, then it is an approximation.  In scikit I use L2 norm.

So we really have a table of similar looking, but rotationally invariant per neighbor distances, `l2norm(Xi - u_i)`, where Xi entries are power.  Power of a signal `x(t)` commonly means `|x(t)|^2`, which I think washes out the sqrt (though I don't believe it even matters up to order, and there is other funny stuff going on in bispectrum).

Hopefully I didn't make too many mistakes or typos.

Reply:
I think just so we don't confuse this heuristic with the other heuristic, I'll make a new name for this guy.

Reply:
renamed `NeighborVarianceClassSelector`